src/librustc_mir/transform/simplify.rs

```diff
@@ -368,21 +368,39 @@ impl<'a, 'tcx> Visitor<'tcx> for DeclMarker<'a, 'tcx> {
             if location.statement_index != block.statements.len() {
                 let stmt = &block.statements[location.statement_index];
 
+                fn can_skip_constant(c: &ty::Const<'tcx>) -> bool {
+                    // Keep assignments from unevaluated constants around, since the
+                    // evaluation may report errors, even if the use of the constant
+                    // is dead code.
+                    !matches!(c.val, ty::ConstKind::Unevaluated(..))
+                }
+
+                fn can_skip_operand(o: &Operand<'_>) -> bool {
+                    match o {
+                        Operand::Copy(p) | Operand::Move(p) => !p.is_indirect(),
```

Comment:
why can't indirect reads get eliminated?

Reply:
They may have been a dereference of a raw pointer.

Reply:
But if it's just a read we should be able to eliminate it. it's not like it's a volatile read that must happen.

Reply:
I can't think of any reason why they couldn't be. I was just being conservative. Compiler bootstrap and tests don't seem to mind though.